uplink/storage/streams/store.go

```diff
@@ -214,39 +219,42 @@ func (s *streamStore) upload(ctx context.Context, path Path, pathCipher storj.Ci
 			beginSegment := &metainfo.BeginSegmentParams{
 				MaxOrderLimit: s.maxEncryptedSegmentSize,
 				Position: storj.SegmentPosition{
-					Index: int32(currentSegment),
+					Index: int32(totalSegments),
 				},
 			}
 
 			var responses []metainfo.BatchResponse
-			if currentSegment == 0 {
+			if totalSegments == 0 {
 				responses, err = s.metainfo.Batch(ctx, beginObjectReq, beginSegment)
 				if err != nil {
-					return Meta{}, currentSegment, streamID, err
+					return Meta{}, err
 				}
 				objResponse, err := responses[0].BeginObject()
 				if err != nil {
-					return Meta{}, currentSegment, streamID, err
+					return Meta{}, err
 				}
 				streamID = objResponse.StreamID
 			} else {
 				beginSegment.StreamID = streamID
 				responses, err = s.metainfo.Batch(ctx, lastCommitSegmentReq, beginSegment)
 				if err != nil {
-					return Meta{}, currentSegment, streamID, err
+					return Meta{}, err
 				}
 			}
+
+			totalSegments++
```

Comment:
Why are we incrementing `totalSegments` that early?

Reply:
Because at this point the segment is created in the satellite and if there is an error we have to clean it up.

Reply:
Are you sure? As far as I can see the pointer is written to the metainfo db on CommitSegment/Object, not on BeginSegment/Object.

Reply:
That's what I understood in the previous meeting, but probably I was confused with the inline segments which don't require commit.

But incrementing at the end of the loop can cause the same problem because the commit of the previous segment happens in the next iteration, so we have to look for another way in how to deal with it.